File: web/js/device.js
// Protocol client on top of a transport: matches replies to requests by seq, delivers frames.
import * as P from './protocol.js';

export class DeviceError extends Error {}

export class Device extends EventTarget {
  constructor(transport) {
    super();
    this.t = transport;
    this.splitter = new P.FrameSplitter();
    this.seq = 0;
    this.pending = new Map();   // seq -> {resolve, reject, msgs, timer}
    this.badFrames = 0;
    this.latest = new Map();    // coalescing key -> {type, body, waiters}
    this.inflight = new Set();
    this.connected = false;
  }

  async open() {
    this.t.onbytes = (b) => this._onBytes(b);
    this.t.ondisconnect = (reason) => this._lost(reason);
    await this.t.open();
    this.connected = true;
  }

  async close() {
    this.connected = false;
    await this.t.close();
    this._failAll('closed');
  }

  _lost(reason) {
    if (!this.connected) return;
    this.connected = false;
    this._failAll(reason);
    this.dispatchEvent(new CustomEvent('disconnect', { detail: reason }));
  }

  _failAll(reason) {
    for (const p of this.pending.values()) { clearTimeout(p.timer); p.reject(new DeviceError(reason)); }
    this.pending.clear();
  }

  _onBytes(chunk) {
    for (const f of this.splitter.push(chunk)) {
      let m;
      try { m = P.decode(f); } catch { this.badFrames++; continue; }
      if (m.type === P.FRAME || m.type === P.ROLL) {
        let frame;
        try { frame = P.parseFrame(m.body); } catch { this.badFrames++; continue; }
        this.dispatchEvent(new CustomEvent(m.type === P.ROLL ? 'roll' : 'frame', { detail: frame }));
      } else if (m.type === P.LOG) {
        console.info('device:', new TextDecoder().decode(m.body));
      } else {
        const p = this.pending.get(m.seq);
        if (!p) continue;
        p.msgs.push(m);
        if (m.type !== P.TABLE) {  // TABLEs precede the final ACK
          clearTimeout(p.timer);
          this.pending.delete(m.seq);
          p.resolve(p.msgs);
        }
      }
    }
  }

  request(type, body = new Uint8Array(0), timeout = 2000) {
    if (!this.connected) return Promise.reject(new DeviceError('not connected'));
    do { this.seq = (this.seq % 255) + 1; } while (this.pending.has(this.seq));
    const seq = this.seq;
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(seq);
        reject(new DeviceError(`timeout waiting for reply to 0x${type.toString(16)}`));
      }, timeout);
      this.pending.set(seq, { resolve, reject, msgs: [], timer });
      this.t.write(P.encode(type, seq, body)).catch((e) => {
        clearTimeout(timer);
        this.pending.delete(seq);
        reject(e);
      });
    });
  }

  async command(type, body, timeout) {
    const msgs = await this.request(type, body, timeout);
    const last = msgs[msgs.length - 1];
    if (last.type !== P.ACK) throw new DeviceError(`expected ACK, got 0x${last.type.toString(16)}`);
    if (last.body[0]) throw new DeviceError(P.ACK_NAMES[last.body[0]] ?? `status ${last.body[0]}`);
    return msgs.slice(0, -1);
  }

  /** Like command(), but while one with the same key is in flight only the newest waiting
   * request is kept. Dragging a slider sends at most one command per round trip. */
  coalesce(key, type, body) {
    return new Promise((resolve, reject) => {
      const prev = this.latest.get(key);
      const waiters = prev ? prev.waiters : [];
      waiters.push({ resolve, reject });
      this.latest.set(key, { type, body, waiters });
      if (!this.inflight.has(key)) this._drain(key);
    });
  }

  /** coalesce() for a sequence of commands: `fn` runs with the device and is awaited. */
  coalesceTask(key, fn) { return this.coalesce(key, null, fn); }

  async _drain(key) {
    this.inflight.add(key);
    while (this.latest.has(key)) {
      const { type, body, waiters } = this.latest.get(key);
      this.latest.delete(key);
      try {
        if (type === null) await body(this); else await this.command(type, body);
        waiters.forEach((w) => w.resolve());
      } catch (e) {
        waiters.forEach((w) => w.reject(e));
      }
    }
    this.inflight.delete(key);
  }

  // ------------------------------------------------------------------ API

  async hello() {
    const [m] = await this.request(P.HELLO);
    if (m.type !== P.INFO) throw new DeviceError('bad HELLO reply');
    return P.parseInfo(m.body);
  }

  async state() {
    const [m] = await this.request(P.GET_STATE);
    if (m.type !== P.STATE) throw new DeviceError('bad GET_STATE reply');
    return P.parseState(m.body);
  }

  async tables() {
    const out = {};
    for (const m of await this.command(P.GET_TABLES)) {
      if (m.type === P.TABLE) { const t = P.parseTable(m.body); out[t.id] = t.value; }
    }
    return out;
  }

  async storeRead() {
    const [m] = await this.request(P.STORE_READ);
    if (m.type !== P.STORE_DATA) throw new DeviceError('bad STORE_READ reply');
    const n = m.body[0] | m.body[1] << 8;
    return m.body.slice(2, 2 + n);
  }

  async storeWrite(blob) { await this.command(P.STORE_WRITE, blob); }

  /** Install an APP1 image over USB (firmware >= 0.6): stage, verify, commit. The device
   * resets into the new firmware right after the final ACK, so a disconnect follows. */
  async fwUpdate(image, progress = () => {}) {
    const crc = P.crc32(image);
    await this.command(P.SET_ACQ, P.body.acq(P.ACQ_STOP, 100));
    progress(0, 'erasing');
    await this.command(P.FW_BEGIN, P.body.fwRange(image.length, crc), 10000);
    for (let off = 0; off < image.length; off += P.FW_CHUNK) {
      await this.command(P.FW_DATA, P.body.fwData(off, image.subarray(off, off + P.FW_CHUNK)), 5000);
      progress(Math.min(off + P.FW_CHUNK, image.length) / image.length, 'writing');
    }
    await this.command(P.FW_COMMIT, P.body.fwRange(image.length, crc), 5000);
  }

  setChannel(ch, range, coupling, offset) {
    return this.coalesce(`ch${ch}`, P.SET_CHANNEL, P.body.channel(ch, range, coupling, offset));
  }

  setRate(hz) { return this.coalesce('rate', P.SET_TIMEBASE, P.body.timebase(hz)); }

  setTrigger(source, kind, level, width) {
    return this.coalesce('trig', P.SET_TRIGGER, P.body.trigger(source, kind, level, width));
  }

  setAcq(mode, autoMs) { return this.coalesce('acq', P.SET_ACQ, P.body.acq(mode, autoMs)); }

  setGen(mode, freq, duty) { return this.coalesce('gen', P.SET_GEN, P.body.gen(mode, freq, duty)); }

  /** Analog output: uploads the DAC table, then starts it (one coalesced step, so the table
   * always arrives before the SET_GEN that uses it). */
  setGenWave(codes, freq) {
    return this.coalesceTask('gen', async (d) => {
      await d.command(P.SET_WAVE, P.body.wave(codes));
      await d.command(P.SET_GEN, P.body.gen(P.GEN_ANALOG, freq, 50));
    });
  }

  setSystem(backlight, beep) { return this.coalesce('sys', P.SET_SYSTEM, P.body.system(backlight, beep)); }
}

File: tools/dsoq/device.py
"""Async client for the DSO Quad web-control firmware."""
import asyncio
import glob
import struct

import serial_asyncio_fast

from protocol import (ACK, ACK_NAMES, FRAME, GET_STATE, GET_TABLES, HELLO, INFO, LOG, PARAM_SET, PING,
                      PONG, REBOOT, REG_GET, REG_SET, REG_VALUE, SET_ACQ, SET_CHANNEL, SET_GEN,
                      SET_SYSTEM, SET_TIMEBASE, SET_TRIGGER, STATE, TABLE, Frame, State, decode,
                      encode, parse_table)


def find_port():
    ports = sorted(glob.glob('/dev/serial/by-id/*DSO_Quad*'))
    if not ports:
        raise SystemExit('no DSO Quad serial port found')
    return ports[0]


class DeviceError(Exception):
    pass


class Device:
    """One connection. Replies are matched to requests by seq; frames go to a queue."""

    def __init__(self):
        self.reader = self.writer = None
        self.frames = asyncio.Queue(maxsize=8)
        self.bad_frames = 0
        self._seq = 0
        self._pending = {}      # seq -> (future, collected messages)
        self._task = None
        self.raw_log = None     # file object: every byte received is appended (for `record`)

    async def open(self, port=None):
        self.reader, self.writer = await serial_asyncio_fast.open_serial_connection(
            url=port or find_port(), baudrate=115200)
        self._task = asyncio.create_task(self._read_loop())
        return self

    async def close(self):
        if self._task:
            self._task.cancel()
        if self.writer:
            self.writer.close()

    # ------------------------------------------------------------------ transport

    async def _read_loop(self):
        while True:
            chunk = await self.reader.readuntil(b'\x00')
            if self.raw_log:
                self.raw_log.write(chunk)
            if len(chunk) <= 1:
                continue
            try:
                mtype, seq, body = decode(chunk[:-1])
            except ValueError:
                self.bad_frames += 1
                continue
            if mtype == FRAME:
                if self.frames.full():
                    self.frames.get_nowait()  # drop the oldest; keep up with the device
                self.frames.put_nowait(Frame.parse(body))
            elif mtype == LOG:
                print('device:', body.decode(errors='replace'))
            elif seq in self._pending:
                fut, msgs = self._pending[seq]
                msgs.append((mtype, body))
                if mtype != TABLE and not fut.done():  # TABLEs precede the final ACK
                    fut.set_result(msgs)

    async def request(self, mtype, body=b'', timeout=2.0):
        self._seq = (self._seq % 255) + 1
        seq = self._seq
        fut = asyncio.get_running_loop().create_future()
        self._pending[seq] = (fut, [])
        self.writer.write(encode(mtype, seq, body))
        try:
            return await asyncio.wait_for(fut, timeout)
        finally:
            del self._pending[seq]

    async def command(self, mtype, body=b''):
        msgs = await self.request(mtype, body)
        mt, reply = msgs[-1]
        if mt != ACK:
            raise DeviceError(f'expected ACK, got {mt:#x}')
        if reply[0]:
            raise DeviceError(f'{ACK_NAMES[reply[0]] if reply[0] < len(ACK_NAMES) else reply[0]}')
        return msgs[:-1]

    # ------------------------------------------------------------------ API

    async def hello(self):
        (mt, b), = await self.request(HELLO)
        assert mt == INFO
        proto, serial = struct.unpack_from('<HI', b)
        return {'proto': proto, 'serial': f'{serial:08X}', 'fw': b[6:].split(b'\0')[0].decode()}

    async def ping(self, data=b'ping'):
        (mt, b), = await self.request(PING, data)
        return mt == PONG and b == data

    async def state(self):
        (mt, b), = await self.request(GET_STATE)
        assert mt == STATE
        return State.parse(b)

    async def tables(self):
        return dict(parse_table(b) for mt, b in await self.command(GET_TABLES) if mt == TABLE)

    async def set_channel(self, ch, range_, coupling, offset):
        await self.command(SET_CHANNEL, bytes([ch, range_, coupling, offset]))

    async def set_rate(self, hz):
        await self.command(SET_TIMEBASE, struct.pack('<I', int(hz)))

    async def set_trigger(self, source, kind, level, width=0):
        await self.command(SET_TRIGGER, struct.pack('<BBBH', source, kind, level, width))

    async def set_acq(self, mode, auto_ms=100):
        await self.command(SET_ACQ, struct.pack('<BH', mode, auto_ms))

    async def set_gen(self, mode, freq_hz=1000, duty=50):
        await self.command(SET_GEN, struct.pack('<BIB', mode, int(freq_hz), duty))

    async def set_system(self, backlight=255, beep=255):
        await self.command(SET_SYSTEM, bytes([backlight, beep]))

    async def reg_set(self, obj, value):
        await self.command(REG_SET, struct.pack('<BI', obj, value))

    async def reg_get(self, kind):
        (mt, b), = await self.request(REG_GET, bytes([kind]))
        assert mt == REG_VALUE
        return struct.unpack('<BI', b)[1]

    async def param_set(self, addr, value):
        await self.command(PARAM_SET, bytes([addr, value]))

    async def reboot(self, to_fallback=False):
        await self.command(REBOOT, bytes([1 if to_fallback else 0]))

    async def next_frame(self, timeout=5.0):
        return await asyncio.wait_for(self.frames.get(), timeout)
